src/pipeline-stages/src/main/scala/Timer.scala

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import org.apache.spark.ml.{Estimator, Model, PipelineStage, Transformer}
+import org.apache.spark.ml.param.{BooleanParam, ParamMap, PipelineStageParam}
+import org.apache.spark.ml.util.{DefaultParamsReadable, DefaultParamsWritable, Identifiable}
+import org.apache.spark.sql.{DataFrame, Dataset}
+import org.apache.spark.sql.types.StructType
+
+import scala.reflect.runtime.universe.{TypeTag, typeTag}
+
+object Timer extends DefaultParamsReadable[Timer]
+
+trait TimerParams extends Wrappable {
+  val stage = new PipelineStageParam(this, "stage", "the stage to time")
+
+  def getStage: PipelineStage = $(stage)
+
+  val logToScala = BooleanParam(this, "logToScala", "Whether to output the time to the scala console", true)
+
+  def getLogToScala: Boolean = $(logToScala)
+
+  def setLogToScala(v: Boolean): this.type = set(logToScala, v)
+
+  val cacheBefore = BooleanParam(this, "cacheBefore", "whether to cache the dataframe before entering the timer", true)
+
+  def getCacheBefore: Boolean = $(cacheBefore)
+
+  def setCacheBefore(v: Boolean): this.type = set(cacheBefore, v)
+
+  val countBefore = BooleanParam(this, "countBefore", "whether to count the dataframe before entering the timer", true)
+
+  def getCountBefore: Boolean = $(countBefore)
+
+  def setCountBefore(v: Boolean): this.type = set(countBefore, v)
+
+  val cacheAfter = BooleanParam(this, "cacheAfter", "whether to cache the dataframe before exiting the timer", true)
+
+  def getCacheAfter: Boolean = $(cacheAfter)
+
+  def setCacheAfter(v: Boolean): this.type = set(cacheAfter, v)
+
+  val countAfter = BooleanParam(this, "countAfter", "whether to count the dataframe before exiting the timer", true)
+
+  def getCountAfter: Boolean = $(countAfter)
+
+  def setCountAfter(v: Boolean): this.type = set(countAfter, v)
+
+  protected def formatTime(t: Long, isTransform: Boolean, count: Option[Long], stage: PipelineStage): String = {
+    val time = {
+      t / 1000000000.0
+    }
+    val action = if (isTransform) "transform" else "predict"
+    val amount = count match {
+      case Some(i) => s"$i rows"
+      case _ => ""
+    }
+    s"${stage} took ${time}s to $action $amount"
+  }
+
+  protected def log(str: String): Unit = {
+    if (str != "") {
+      println(str)
+    }
+  }
+
+}
+
+class Timer(val uid: String) extends Estimator[TimerModel]
```

Comment:
As discussed, please consider only adding an enable/disable option instead of cache/count x before/after